File: frontend/src/services/userService.ts
import { apiClient } from './apiClient';
import { User, WishlistItem } from '../types';

export interface ProfileUpdateRequest {
  fullName: string;
  email: string;
  phone?: string;
  address?: string;
}

export interface ChangePasswordRequest {
  currentPassword: string;
  newPassword: string;
}

export const userService = {
  async getAll(): Promise<User[]> {
    const res = await apiClient.get<User[]>('/users');
    return res.data;
  },

  async updateRole(id: number, role: 'ADMIN' | 'CUSTOMER'): Promise<User> {
    const res = await apiClient.patch<User>('/users/' + id + '/role', null, { params: { value: role } });
    return res.data;
  },
  async getMe(): Promise<User> {
    const res = await apiClient.get<User>('/users/me');
    return res.data;
  },

  async updateMe(data: ProfileUpdateRequest): Promise<User> {
    const res = await apiClient.put<User>('/users/me', data);
    return res.data;
  },

  async changePassword(data: ChangePasswordRequest): Promise<void> {
    await apiClient.put('/users/me/password', data);
  },

  async getWishlist(): Promise<WishlistItem[]> {
    const res = await apiClient.get<WishlistItem[]>('/users/me/wishlist');
    return res.data;
  },

  async addToWishlist(bookId: number): Promise<WishlistItem> {
    const res = await apiClient.post<WishlistItem>(`/users/me/wishlist/${bookId}`);
    return res.data;
  },

  async removeFromWishlist(bookId: number): Promise<void> {
    await apiClient.delete(`/users/me/wishlist/${bookId}`);
  },
};

File: backend/src/main/java/com/bookstore/user/service/UserService.java
package com.bookstore.user.service;

import com.bookstore.user.dto.ChangePasswordRequest;
import com.bookstore.user.dto.ProfileUpdateRequest;
import com.bookstore.user.entity.User;
import com.bookstore.user.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserService {
  private final UserRepository users;
  private final PasswordEncoder passwordEncoder;

  public UserService(UserRepository users, PasswordEncoder passwordEncoder) {
    this.users = users;
    this.passwordEncoder = passwordEncoder;
  }

  @Transactional(readOnly = true)
  public User getByUsername(String username) {
    return users.findByUsername(username).orElseThrow();
  }

  @Transactional
  public User updateProfile(String username, ProfileUpdateRequest request) {
    var user = getByUsername(username);
    var emailChanged = !user.getEmail().equalsIgnoreCase(request.email());
    if (emailChanged && users.existsByEmailIgnoreCase(request.email())) {
      throw new IllegalArgumentException("Email đã tồn tại");
    }
    user.setFullName(request.fullName());
    user.setEmail(request.email());
    user.setPhone(request.phone());
    user.setAddress(request.address());
    return users.save(user);
  }


  @Transactional
  public User updateRole(String currentUsername, Long userId, String role) {
    var normalizedRole = role == null ? "" : role.trim().toUpperCase();
    if (!normalizedRole.equals("ADMIN") && !normalizedRole.equals("CUSTOMER")) {
      throw new IllegalArgumentException("Role must be ADMIN or CUSTOMER");
    }

    var currentUser = getByUsername(currentUsername);
    var user = users.findById(userId).orElseThrow();
    if (currentUser.getId().equals(user.getId())) {
      throw new IllegalArgumentException("You cannot change your own role");
    }

    user.setRole(normalizedRole);
    return users.save(user);
  }
  @Transactional
  public void changePassword(String username, ChangePasswordRequest request) {
    var user = getByUsername(username);
    if (!passwordEncoder.matches(request.currentPassword(), user.getPassword())) {
      throw new IllegalArgumentException("Mật khẩu hiện tại không đúng");
    }
    user.setPassword(passwordEncoder.encode(request.newPassword()));
    users.save(user);
  }
}
